tests/MongoDB.Driver.Tests/Specifications/change-streams/ChangeStreamTestRunner.cs

```diff
@@ -0,0 +1,414 @@
+﻿/* Copyright 2018-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using FluentAssertions;
+using MongoDB.Bson;
+using MongoDB.Bson.TestHelpers.JsonDrivenTests;
+using MongoDB.Bson.TestHelpers.XunitExtensions;
+using MongoDB.Driver.Core;
+using MongoDB.Driver.Core.Clusters;
+using MongoDB.Driver.Core.Events;
+using MongoDB.Driver.Core.TestHelpers.JsonDrivenTests;
+using MongoDB.Driver.Core.TestHelpers.XunitExtensions;
+using MongoDB.Driver.TestHelpers;
+using MongoDB.Driver.Tests.Specifications.crud;
+using Xunit;
+
+namespace MongoDB.Driver.Tests.Specifications.change_streams
+{
+    public class ChangeStreamTestRunner
+    {
+        // private fields
+        private string _databaseName;
+        private string _database2Name;
+        private string _collectionName;
+        private string _collection2Name;
+
+        // public methods
+        [SkippableTheory]
+        [ClassData(typeof(TestCaseFactory))]
+        public void Run(JsonDrivenTestCase testCase)
+        {
+            Run(testCase.Shared, testCase.Test);
+        }
+
+        // private methods
+        private void Run(BsonDocument shared, BsonDocument test)
+        {
+            JsonDrivenHelper.EnsureAllFieldsAreValid(shared, "_path", "database_name", "database2_name", "collection_name", "collection2_name", "tests");
+            JsonDrivenHelper.EnsureAllFieldsAreValid(test, "description", "minServerVersion", "topology", "target", "changeStreamPipeline", "changeStreamOptions", "operations", "expectations", "result");
+
+            if (test.Contains("minServerVersion"))
+            {
+                var minServerVersion = test["minServerVersion"].AsString;
+                RequireServer.Check().VersionGreaterThanOrEqualTo(minServerVersion);
+            }
+
+            if (test.Contains("topology"))
+            {
+                var clusterTypes = MapTopologyToClusterTypes(test["topology"].AsBsonArray);
+                RequireServer.Check().ClusterTypes(clusterTypes);
+            }
+
+            _databaseName = shared["database_name"].AsString;
+            _database2Name = shared["database2_name"].AsString;
```

Comment:
FYI _database2Name and _collection2Name don't actually seem to be used by the JSON files.